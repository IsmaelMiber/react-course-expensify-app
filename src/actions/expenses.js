import uuid from 'uuid';

//ADD_EXPENSE
export const addExpense = ({description = '', note = '', amount = 0, createAt = 0} = {}) => (
    {
        type: "ADD_EXPENSE",
        expense: {
            id: uuid(),
            amount,
            description,
            note,
            createAt,
        }
    }
);

//REMOVE_EXPENSE
export const removeExpense = (idOfTargetExpense) => (
    {
        type: "REMOVE_EXPENSE",
        idOfTargetExpense,
    }
);

//EDIT_EXPENSE
export const editExpense = (id, updates) => (
    {
        type: "EDIT_EXPENSE",
        id,
        updates,
    }
);