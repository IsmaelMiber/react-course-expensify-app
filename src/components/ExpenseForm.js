import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
    state = {
        description: this.props.expense ? this.props.expense.description : '',
        note: this.props.expense ? this.props.expense.note : '',
        amount: this.props.expense ? (this.props.expense.amount / 100).toString() : '',
        createAt: this.props.expense ? moment(this.props.expense.createAt) : moment(),
        calendarFocused: false,
        error: '',
    };

    onDescriptionChange = (e) => {
        const description = e.target.value.trim();
        this.setState(() => ({description}));
    }

    onNoteChange = (e) => {
        const note = e.target.value.trim();
        this.setState(() => ({note}));
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        if( !amount || amount.match(/^\d{1,}(\.\d{0,2})?$/) ) this.setState(() => ({amount}));
    }

    onDateChange = (createAt) => {
        if(createAt) this.setState(() => ({createAt}));
    }

    onFocusChange = ({focused}) => {
        this.setState(()=>({calendarFocused: focused}));
    }

    onSubmit = (e) => {
        e.preventDefault();

        if(!this.state.description || !this.state.amount) {
            this.setState(()=>({error: 'Please, provide description & amount'}));
        } else {
            this.setState(()=>({error: ''}));
            this.props.onSubmit({
                description: this.state.description,
                note: this.state.note,
                amount: parseFloat(this.state.amount, 10) * 100,
                createAt: this.state.createAt.valueOf(),
            })
        }
    }

    render() {
        return (
            <div>
                {this.state.error ? <p>{this.state.error}</p> : this.state.error}

                <form onSubmit={this.onSubmit}>
                    <input
                    type="text"
                    placeholder="Description"
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                    />

                    <input
                    type="text"
                    placeholder="Amount"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                    />

                    <SingleDatePicker
                    date={this.state.createAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    />

                    <textarea
                    placeholder="Add a note for your expense (optional)"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                    >
                    </textarea>

                    <button>Submit</button>
                </form>
            </div>
        );
    }
}