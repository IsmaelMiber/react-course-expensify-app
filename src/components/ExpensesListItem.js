import React from 'react';
import { Link } from 'react-router-dom';

const ExpensesListItem = (props) => (
    <div>
        <p><Link to={`/edit/${props.id}`}>description: {props.description}</Link></p>
        <p>amount: {props.amount}</p>
        <p>createdAt: {props.createAt}</p>
        <hr />
    </div>
);

export default ExpensesListItem;