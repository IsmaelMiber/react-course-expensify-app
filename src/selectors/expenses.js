import moment from 'moment';

export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter( ({ createAt, description }) => {
            const createAtMoment = moment(createAt);
            const startDateMatch = startDate ? startDate.isSameOrBefore(createAtMoment, 'day') : true;
            const endDateMatch = endDate ? endDate.isSameOrAfter(createAtMoment, 'day') : true;
            const textMatch = description.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;
    }).sort( (a, b) => {
        if(sortBy === 'date') {
            return a.createAt < b.createAt ? 1 : -1;
        } else {
            return a.amount < b.amount ? 1 : -1;
        }
    });
};