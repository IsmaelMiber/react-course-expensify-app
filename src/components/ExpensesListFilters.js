import React from 'react';
import { connect } from 'react-redux';
import * as filterActions from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpensesListFilters extends React.Component {

    state = {
        focusedInput: null,
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange = (focusedInput) => this.setState(()=>({focusedInput}));

    onSortByChange = (e) => e.target.value === 'date' ? this.props.sortByDate() : this.props.sortByAmount();

    onTextChange =  (e) => this.props.setTextFilter(e.target.value);

    render() {
        return (
            <div>
                <input
                    type="text"
                    value={this.props.filters.text}
                    onChange={this.onTextChange}
                 />
                <select
                    value={this.props.filters.sortBy}
                    onChange={this.onSortByChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>

                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.focusedInput}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}

                />
            </div>
        );
    }
}

const mapStateToProps = (state) => (
    {
        filters: state.filters,
    }
);

export default connect(mapStateToProps, filterActions)(ExpensesListFilters);