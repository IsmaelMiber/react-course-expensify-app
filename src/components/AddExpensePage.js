import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        // props.dispatch(addExpense(expense));
        this.props.addExpense(expense);
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                />
            </div>
        );
    };
};

const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);