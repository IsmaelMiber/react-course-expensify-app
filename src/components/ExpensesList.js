import React from 'react';
import { connect } from 'react-redux';
import ExpensesListItem  from './ExpensesListItem';
import selectExpenses from '../selectors/expenses';

export const ExpensesList = (props) => (
    <div>
        <h1>Expenses List</h1>
        { props.expenses.length !== 0 ?
            props.expenses.length > 0 && props.expenses.map(expense => (
            <ExpensesListItem key={expense.id} {...expense}/>
        )) : (
             <p>There is no expenses</p>
             )
        }
    </div>
);

const mapStateToProps = state => (
    {
        expenses: selectExpenses(state.expenses, state.filters),
    }
);

export default connect(mapStateToProps)(ExpensesList);