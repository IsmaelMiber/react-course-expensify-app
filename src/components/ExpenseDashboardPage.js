import React from 'react';
import ExpensesList from './ExpensesList';
import ExpensesListFilters from './ExpensesListFilters';

const ExpenseDashboardPage = () => (
    <div>
        This is my ExpenseDashboardPage component
        <ExpensesListFilters />
        <ExpensesList />
    </div>
);

export default ExpenseDashboardPage;