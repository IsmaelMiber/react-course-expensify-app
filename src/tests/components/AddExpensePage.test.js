import React from 'react';
import { shallow } from 'enzyme';
import { AddExpensePage } from '../../components/AddExpensePage';
import expenses from '../fixtures/expenses';

let addExpense, history, wrapper;

beforeEach(() => {
    addExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<AddExpensePage addExpense={addExpense} history={history} />);
});

test("Should AddExpensePage render correctly", () => {
    expect(wrapper).toMatchSnapshot();
});

test("Should AddExpensePage handle correctly", () => {
    wrapper.find("ExpenseForm").prop("onSubmit")(expenses[0]);
    expect(addExpense).toHaveBeenCalledWith(expenses[0]);
    expect(history.push).toHaveBeenCalledWith('/');
});