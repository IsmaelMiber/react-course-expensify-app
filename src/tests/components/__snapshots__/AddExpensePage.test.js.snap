// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should AddExpensePage render correctly 1`] = `
<div>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
